### ISSUE TITLE
[feature] Allow to render content with block

### ISSUE THREAD

**Issue by blelump** [apotonick/cells#382]
Hi @apotonick ! 

First, consider such piece of code:

```
class Navigation::Cell < Cell::Concept;  end
class Dashboard::Cell < Navigation::Cell; end
```

Now let me explain. My motivations were to build a view based upon OOP inheritance. With such approach, I imagined, I could build cell content based on class type and its superclasses. Well, it turned out it isn't so easy, because I cannot nest one view component into another, because the [renderer][1] structure is flat. It simply looks for a template, renders it and puts into layout, if any is given.

So I tweaked it a bit to let render take a `block` and create nested structures. It looks like below:
```
render(view: :some_part_of_dashboard, layout: :dashboard) do
  render(view: :another_part_of_dashboard)
end
```

In particular, it just joins given view components into one bigger component, which might be further wrapped into a layout (if given, of course). My motivation was to extract common components and build dashboard page (it's quite complex and role dependent) upon what is needed.

Basically, what I'd like to achieve is:
```
def show
  render :component + render :another-component
end
```
And wrap the output it into a layout

-----

These changes were made:
```
def render(options={}, &block)
  options = normalize_options(options)
  render_to_string(options, &block)
end

def render_to_string(options)
  template = find_template(options)
  nested_content = yield if block_given?
  content = render_template(template, options) + nested_content.to_s

  # TODO: allow other (global) layout dirs.
  with_layout(options, content)
end
```

Does it make any sense to you? Perhaps it might be solved in an easier way?

[1]: https://github.com/apotonick/cells/blob/v4.0.4/lib/cell/view_model.rb#L114

**Comment by timoschilling** [comment]
Your code brings the problem that your `block` / `nested_content` is just append to the view.

I will write a change to support `render(options={}, &block)` and `yield` usage in the template.

**Comment by blelump** [comment]
Hi @timoschilling !

Thanks for answering. Yep, my code is probably far from ideal. Could you elaborate your approach?

**Comment by timoschilling** [comment]
I hope I can rewrite many of the rendering things in the near future.

This will maybe work then:
```ruby
class ACell
  def show
    render :foo do
      detail
    end
  end

  def detail
    ...
  end
end
# or
class ACell
  def show
    render :foo do
      AOtherCell.(model).()
    end
  end
end
```
View:
```ruby
<h1><%= name %></h1>
<div class="content">
  <%= yield %>
</div>
```

**Comment by blelump** [comment]
I see, thanks! Your proposal sounds better and I think it should be quite easily affordable with the current API. I am not sure yet (will try it tomorrow), but I think for now it would be enough to pass a block [here](https://github.com/apotonick/cells/blob/master/lib/cell/view_model.rb#L103) and fire `yield` in particular view.

**Comment by blelump** [comment]
Hi @timoschilling ,

I've checked it and it works, however I am not sure this is what I'd expect. I mean with such approach you have smth like:
```
# view.erb
  <%= yield %>

def render_to_string(options, &block)
  # ...
  content = render_template(template, options, &block)
  # ...
end
``` 

which explicitly enforce using `yield` in view, whereas this is not the intention, because I'd like to have these view components to be generic widgets and built more complex layout from them. 

Is it what did you mean?

**Comment by blelump** [comment]
Hi guys,

I've experimented a bit more with cells rendering and it seems, at least with slim, that everything is working. Now, I'm able to do smth like:
```
render view: :main do
  render(view: :content) + render(view: :sidebar)
end
```

or even
```
render view: :main do
  render(view: :content) + concept('smth/sidebar/cell').to_s
end
```

![example](http://i.imgur.com/js6NxMN.png)

It works for Slim and to make it working with ERB and HAML, probably more changes are needed. Essentially I've removed buffer from `template_options_for` and passed a block [here](https://github.com/apotonick/cells/blob/master/lib/cell/view_model.rb#L103). If working with pure Tilt, it plays really smoothly.

**Comment by apotonick** [comment]
What about this?

```ruby
render :layout, locals: { content: render(:content_view) }
```

This will do the trick and avoid any problems about buffers and execution contexts. It's incredibly hard for me to find a solution for the `yield` approach in views which is essentially unnecessary as you can pass layout content as a local, or am I wrong here?

**Comment by timoschilling** [comment]
render :layout do
end

**Comment by timoschilling** [comment]
This should work without any buffer conflicts.
```ruby
render :layout do
  concept(...).()
end
```

**Comment by blelump** [comment]
Hi guys!

The approach I've shown above evaluated, a bit :smile:  . Because I am able to work with blocks, it lets me to do smth like:
```
    render view: :form, layout: :main do |f|
      render view: :show, locals: {f: f}
    end
```

which essentially not only covers the "parent  child" case, but also allows me to inject additional dependencies into child so I am able to create interactions among my components. You guys for sure ever created a GUI program which have had a graphic layout – perhaps I'm too old, because I'm using such dinosaur–like example :smile: . Today I'd recall React probably... Anyway, in any language or GUI library it works the same, I mean you're able to create interactions among components in a way however you like. So any event happened in compontent X, may interact with component Y or Z in any manner.

I am not sure, but perhaps we would be able to do that with @apotonick suggesstion,  e.g. create a `def form @form ||= Formular::Builder.new(...) ; end` , method so each view template is able to access the `form` . I think the same is affordable with @timoschilling use case, where I'd inject `self` as a concept model (and actually I'm doing it right now in many cases) so the "child" cell is able to access parent.

However, I'm still kind of a fan of the 
```
render :smth do
  render :blah + concept('cool/cell')
end
``` 
approach, because I think it is simply easier to understand – at least for me. I mean, you see what goes where. Perhaps it works so nice for me, because I'm on Slim and Formular, so not touching any AcionView core (OK, except `content_tag`) or these are unrelated? 

What do you guys think about it? Right now I'm out of spare resources to investigate the remaining templates (ERB, HAML) and AcionView compontents (e.g `form_for`), but perhaps I could provide a sample app to show how it spins.

**Comment by apotonick** [comment]
I still don't understand why the example I posted earlier wouldn't work? You could override `#render` in your cells and add the syntactical block sugar, but the mechanics are already there via the `:locals` option? 

Instead of `yield` in your layout, you'd do `= content` - isn't that exactly what you want???

**Comment by blelump** [comment]
@apotonick , I think it will work in most cases just fine. Just wanted to say that the _block_ approach seems to be more flexible, however I'm still experimenting with this and trying to find solutions that seem to be a _best fit_ (yep, that's quite subjective though).

Anyway, I can close it, if you think the way it works now is just fine.

**Comment by blelump** [comment]
closing because of no response

**Comment by apotonick** [comment]
For Cells 5, we are simplifying Cells core even more and really rip out anything that's not 99% useful. Passing a block to `render` is really really just a syntactical sugar that can be implemented easily with `:locals`.

BTW, I added some cool functionality in regards to layouts in Cells 4.1: You can now instruct a cell to use a second "layout" cell to render a container (e.g. the page layout). Would that help you?

Thanks for thinking about this, @blelump, I will have another thought if we should provide this feature.

**Comment by blelump** [comment]
Hi @apotonick ,

thanks for your response! Not sure yet about the second layout, but I've seen the context object and `tap` approach. I've played with it a bit, I mean `Comment::Cell.new(comment).().tap { |cell| } ` and it might be useful in some cases, but I recognize it as a "postprocessing" step, since the `Comment::Cell` is already rendered here. I've tweaked it a bit to be able to render nested, eg:
```
    content = Main::Cell.call(ctxt, controller: self).tap do |parent|
      parent.show do
        concept("documents/cell",
                cell_ctxt(with: {blah: op.contract}))
      end
    end

    render text: content, layout: true
```

and main_cell:
```
class Main::Cell < Isa::Cell::Concept
  attr_reader :result

  def to_s
    result
  end

  def show(&block)
    @result = render view: :main, &block
  end
end
```

You may wonder wtf. 

What I want to achieve here is to separate responsibilities of cells. At first (a couple weeks ago) my intention was to render a "super cell" which also renders all nested cells (if exist). Now I've changed the approach because it mixins the concerns, e.g a subcell should be aware only of the context (an arbitrary object) you put into it. It doesn't need to know who is its parent. However, the parent cell is responsible of rendering its children within a layout, no matter whether its a horizontal, vertical or a mix of these two.

Besides, I'd be really grateful if you'd explain [this](https://github.com/apotonick/cells/blob/master/test/context_test.rb#L22). I mean what should be the result of calling `cell.cell("context_test/child", "")` ? Such `child` cell is nested in parent? Nontheless, the API here is much better, because it is simple `smth.cell('blah')`, however I wonder how it's supposed to work.

**Comment by blelump** [comment]
@apotonick ,  OK, I've checked the new layout feature you've metnioned above. I've set it up as shown in the [test](https://github.com/apotonick/cells/blob/master/test/layout_test.rb#L60) and it spins fine. :fire: 

There's one caveat, however, because what if you have smth like `render(view: :content) + render(view: :sidebar)`? The `render_to_string` from [external](https://github.com/apotonick/cells/blob/master/lib/cell/layout.rb#L23) module is used any time render is performed and hence, the result is that each render is wrapped with layout cell, which is not expected behaviour. If you don't get what I mean, please look on the picture [here](https://github.com/apotonick/cells/issues/382#issuecomment-196024098).

**Comment by apotonick** [comment]
Very very good point, @blelump about the `#render_to_string`. Let me think about this real quick.

**Comment by apotonick** [comment]
Maybe the `External` module shouldn't override `render_to_string` but `show`, as a convention?

**Comment by blelump** [comment]
> Maybe the External module shouldn't override render_to_string but show, as a convention?

@apotonick ,

yeah, `show` output wrapped into a layout sounds good :+1: .  However, I think there should be an intermediate interface that would allow one to use any present method (`show`, `present` or whatever). The point here is that `layout.new(model, context: context, content: content)` needs to get appropriate `content`. 

Note that [`ViewModel#to_s`](https://github.com/apotonick/cells/blob/master/lib/cell/view_model.rb#L131) is overidden with `call`, so why not to just override call method? e.g:

```
module Layout
  def call(*args)
     content = super
     layout.new(model, context: context, content: content)
  end
end
```

Would it work? Sorry, but didn't have time to test it.

**Comment by apotonick** [comment]
Yeah, it's the only way to hook into the "global" render.

**Comment by apotonick** [comment]
If you want, @blelump, you can test the above commit. It works great for me!

**Comment by blelump** [comment]
Hi @apotonick , 

I've just pulled the latest commit from master and it seems it works fine! :fire: :fireworks:
